papers/ice_nucleation_2024/perfect_model_J.jl

```diff
@@ -0,0 +1,270 @@
+
+    # Create prior
+    prior = create_prior(FT, IN_mode)
+
+    if IN_mode == "ABDINM"
+        Nₐ = FT(8e6) # according to fig 2 panel b deposition nucleation experiment
+        Nₗ = FT(0)
+        Nᵢ = FT(0)
+        r₀ = FT(0.5e-6)
+        p₀ = FT(987.018 * 1e2)
+        T₀ = FT(212.978)
+
+        C_v = FT(10.8509 * 1e-6)
+        qᵥ = ϵₘ / (ϵₘ - 1 + 1 / C_v)
+        qₗ = FT(0)
+        qᵢ = FT(0)
+
+        q = TD.PhasePartition.(qᵥ + qₗ + qᵢ, qₗ, qᵢ)
+        Rₐ = TD.gas_constant_air(tps, q)
+        eₛ = TD.saturation_vapor_pressure(tps, T₀, TD.Liquid())
+        e = eᵥ(qᵥ, p₀, Rₐ, R_v)
+        Sₗ = FT(e / eₛ)
+        IC = [Sₗ, p₀, T₀, qᵥ, qₗ, qᵢ, Nₐ, Nₗ, Nᵢ, FT(0)]
+
+        dep_nucleation = "ABDINM"
+        deposition_growth = "Deposition"
+        size_distribution = "Monodisperse"
+
+        params = (;
+            const_dt,
+            w,
+            dep_nucleation,
+            deposition_growth,
+            size_distribution,
+        )
+
+        coeff_true = [FT(27.551), FT(-2.2209)]
+
+    elseif IN_mode == "ABIFM"
+        Nₐ = FT(0)
+        Nₗ = FT(8e6)
+        Nᵢ = FT(0)
+        r₀ = FT(0.5e-6)                 # droplets maybe bigger?
+        p₀ = FT(987.018 * 1e2)
+        T₀ = FT(212.978)
+
+        qₗ = FT(Nₗ * 4 / 3 * FT(π) * r₀^3 * ρₗ / FT(1.2)) # 1.2 should be ρₐ
+        C_l = FT(qₗ / ((1 - qₗ) * ϵₘ + qₗ))  # concentration/mol fraction of liquid
+        C_v = FT(10.8509 * 1e-6 - C_l)     # concentration/mol fraction of vapor
+        qᵥ = ϵₘ / (ϵₘ - 1 + 1 / C_v)
+        qᵢ = FT(0)
+
+        q = TD.PhasePartition.(qᵥ + qₗ + qᵢ, qₗ, qᵢ)
+        Rₐ = TD.gas_constant_air(tps, q)
+        eₛ = TD.saturation_vapor_pressure(tps, T₀, TD.Liquid())
+        e = eᵥ(qᵥ, p₀, Rₐ, R_v)
+        Sₗ = FT(e / eₛ)
+        IC = [Sₗ, p₀, T₀, qᵥ, qₗ, qᵢ, Nₐ, Nₗ, Nᵢ, FT(0)]
+        Sᵢ = ξ(tps, T₀) * Sₗ
+
+        heterogeneous = "ABIFM"
+        condensation_growth = "Condensation"
+        deposition_growth = "Deposition"
+        size_distribution = "Monodisperse"
+
+        params = (;
+            const_dt,
+            w,
+            heterogeneous,
+            condensation_growth,
+            deposition_growth,
+            size_distribution,
+        )
+
+        coeff_true = [FT(54.58834), FT(-10.54758)]
+    end
+
+    n_samples = 10
+    G_truth = run_model(params, coeff_true, IN_mode, FT, IC)     # ICNC from running parcel w default values
+    y_truth = zeros(length(G_truth), n_samples) # where noisy ICNC will be stored
+
+    dim_output = length(G_truth)
+    Γ = 0.005 * LinearAlgebra.I * (maximum(G_truth) - minimum(G_truth))
+    noise_dist = Distributions.MvNormal(zeros(dim_output), Γ)
+    y_truth = G_truth .+ rand(noise_dist)
+
+    # Generate initial ensemble and set up EKI
+    N_ensemble = 10     # runs N_ensemble trials per iteration
+    N_iterations = 15    # number of iterations the inverse problem goes through
+    initial_ensemble = EKP.construct_initial_ensemble(rng, prior, N_ensemble)
```

Comment:
todo - why Float32 doesn't work?

Reply:
`initial_ensemble`, `y_truth`, and `Γ` are Matrix{Float64}, Matrix{Float64}, and UniformScaling{64} respectively (`y_truth` being dependent on `Γ`). `Γ` is some noise multiplied by the identity matrix from the LinearAlgebra package, `I`, which is of the type UniformScaling. I can't use a Matrix{Float64} for the identity matrix because then `EKP.EnsembleKalmanProcess` gets confused and breaks

Reply:
You should be able to use FT32 or FT64 as the FT is inferred from the inputs consistently. 

could you say more about this error with `I` vs `Matrix`? I haven't come across this.

Reply:
Yeah but when I use `FT32`, I get the following error:

<img width="1008" alt="Screenshot 2024-04-05 at 1 39 36 PM" src="https://github.com/CliMA/CloudMicrophysics.jl/assets/121587057/06c8483b-e98e-47e1-bde2-f0f401376bb6">

So it seems like` Matrix{Float32}` doesn't count as an `AbstractMatrix{FT}`? And if I further use `identity_matrix = Matrix{FT}(LinearAlgebra.I, dim_output, dim_output)` instead of using the built in identity matrix `I`, I get an error that looks like this:

<img width="1004" alt="Screenshot 2024-04-05 at 1 51 48 PM" src="https://github.com/CliMA/CloudMicrophysics.jl/assets/121587057/c68f2445-2375-41c4-b4d9-cd5e76e41a9e">

Reply:
These errors are both coming from a mismatch where the `update_ensemble!` or `EnsembleKalmanProcess(...)` is being applied to a function with a bunch of FT64 arguments and one FT32 argument. 
The way it's set up is that either everything needs to be one type or the other. for example you would need the following to use FT32 
```
 EKI_obj = EKP.EnsembleKalmanProcess(
        initial_ensemble, # must contain FT32 data
        y_truth, # must be a FT32 vector
        Γ, # must have FT32 entries
        EKP.Inversion();
        rng = rng,
    )
```

Reply:
Happy to chat further about this if the FloatType is important to preserve!

Reply:
Yeah I just realized that even though I defined everything as `FT`, calling `length(some_array)` returns a `Float64` specifically. Fixing it now. Thanks Ollie!